Change ARRAY_INTERSECT to use TypedSet

yingsu00 opened the pull request:
Resolves:
https://github.com/prestodb/presto/issues/11493

This commit improves the performance comparing to the original sort-merge
solution. The gain is from 2.3x to 11x for different data types.

before

```
Benchmark                                        (name)   (type)  Mode  Cnt    Score   Error  Units
BenchmarkArrayIntersect.arrayIntersect  array_intersect   BIGINT  avgt   20  193.023 ± 2.840  ns/op
BenchmarkArrayIntersect.arrayIntersect  array_intersect  VARCHAR  avgt   20  336.516 ± 5.301  ns/op
BenchmarkArrayIntersect.arrayIntersect  array_intersect   DOUBLE  avgt   20  219.812 ± 6.615  ns/op
BenchmarkArrayIntersect.arrayIntersect  array_intersect  BOOLEAN  avgt   20   23.913 ± 0.553  ns/op
```

After

```
Benchmark                                (type)  Mode  Cnt    Score   Error  Units
BenchmarkArrayIntersect.arrayIntersect   BIGINT  avgt   20   30.175 ± 5.077  ns/op
BenchmarkArrayIntersect.arrayIntersect  VARCHAR  avgt   20  146.365 ± 5.405  ns/op
BenchmarkArrayIntersect.arrayIntersect   DOUBLE  avgt   20   63.752 ± 5.329  ns/op
BenchmarkArrayIntersect.arrayIntersect  BOOLEAN  avgt   20    2.133 ± 0.243  ns/op
```

findepi commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java:
```diff
@@ -48,98 +42,266 @@ public ArrayIntersectFunction(@TypeParameter("E") Type elementType)
+        return blockBuilder.getRegion(blockBuilder.getPositionCount() - intersectCount, intersectCount);
     }
+
+//    private static IntComparator intBlockCompare(Type type, Block block)
```
nit: you probably want to remove this

dain commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java:
```diff
@@ -48,98 +42,266 @@ public ArrayIntersectFunction(@TypeParameter("E") Type elementType)
+            }
+        }
+
+        return blockBuilder.getRegion(blockBuilder.getPositionCount() - intersectCount, intersectCount);
```
I believe that `intersectTypedSet` already contains the output data, so maybe modify `TypedSet` to return `elementBlock.build`.

yingsu00 commented:
I did some JMH benchmarks on the two possible changes to TypedSet:

	1) Instead of building an elementBlock inside of TypdeSet, we can directly pass in the block and only build a hashtable of block positions. This way we avoid one data copy and Block building, and also removed the 4MB limit on the elementBlock Size. However as @haozhun pointed out this may have memory GC issue for aggregatoin functions like MultimapAggregationFunction , because this way we have to hold the input block longer until we're done with the TypedSet, while the other array/map functions don't matter that much because the life cycle of the input blocks and TypedSet are not muh different. But since there're only one aggregation function, this can probably work.
	Another change we need to make this work is to be able to add more than one block to the TypedSet. There are a few use cases(listed in the table below) requires this, e.g. ArrayExceptFunction. This is doable, we just need to keep track of the added blocks inside of TypedSet, and maybe use some additional internal hashtable to find which block the value is in. There is a space penalty for this of course, but because we don't need to build the internal elementBlock the space requirement should not be more than existing implementation.
	
	2) There're a number of use cases the internal elementBlock can be DIRECTLY returned and used as results. Such cases are also listed in the table below. Theoretically this would save some cost of building the a separate result blockBuilder in caller(e.g. distinctElementBlockBuilder in ArrayDistinctFunction), however in the JMH benchmarks I'm not seeing any obvious performance gains, except that this should save memory usage any ways. If we do 1) we can also make TypdeSet to return a result block based on an array of positions, but again this doesn't show any performance gain/loss comparing to using a separate result block builder in the caller.


This table lists all usages of TypdedSet. We can see 5 of them requires the ability to add more than one Blocks so this feature is needed. Also there're 6 callers that can directly use the current TypedSet internal elementBlock as result. However if we use approach 1) and just pass in the input block, the benefit of saving memory usage by just returning the elementBlock directly goes away.

```
  | Aggregation Function | Input Blocks | Handle > 1 blocks add | Can directly use the elementBlock as output
-- | -- | -- | -- | --
MultimapAggregationFunction | Y | 1 | N | Y
ArrayDistinctFunction | N | 1 | N | Y
ArrayExceptFunction | N | 2 | Y | N
ArrayIntersectFunction (Using hashtable solution) | N | 2 | N | Y
ArrayUnionFunction | N | 2 | Y | Y
MapConcatFunction | N | n | Y | Y
MapFromEntriesFunction | N | 1   N SingleRowBlock | Y | Y (for keyBlock)
MapToMapCast | N | 1 | N | Y(need to apply valueProcessFunction.invokeExact)
mapDotProduct | N | 2 | N | N
MultimapFromEntriesFunction | N | 1 | N | N
```
Without knowing the existence of JsonUtil#HashTable, I implemented a class NoCopyTypedSet and did a few benchmarks on array_distinct. The logic is similar to JsonUtil#HashTable, except NoCopyTypedSet handles nulls while JsonUtil#HashTable doesnt. If we use JsonUtil#HashTable directly to implement functions like array_distinct it would give wrong results on 0 and null.  As a result NoCopyTypedSet is slightly slower than HashTable for about 10%. The tests were done on arrays with 10000 varchar values.

The first test shows directly returning elementBlock doesn't have much gain in performance:

TypedSet, caller build the result Block in a separrate block builder:
BenchmarkArrayDistinct.arrayDistinct      array_distinct  avgt   20  35.440 ± 2.055  ns/op

TypedSet, directly return elementBlock
BenchmarkArrayDistinct.arrayDistinct  array_distinct  avgt   20  34.890 ± 0.916  ns/op


The next test using NoCopyTypedSet compares 1) return a block in NoCopyTypedSet .getBlock() vs 2)the caller uses a separate block builder:

NoCopyTypedSet  return the result Block in NoCopyTypedSet.getBlock() by calling Block.getPositions():
BenchmarkArrayDistinct.arrayDistinct      array_distinct  avgt   20  29.871 ± 2.498  ns/op

NoCopyTYpedSet, caller use separate blockBuilder
BenchmarkArrayDistinct.arrayDistinct      array_distinct  avgt   20  28.951 ± 1.027  ns/op


Again there's no obvious difference. NoCopyTypedSet  is faster than TypedSet because I changed the internal hashtable to int[] instead of IntArrayList.

The JsonUtil#HashTable results;
JsonUtil#HashTable,  return the result Block in NoCopyTypedSet.getBlock()
BenchmarkArrayDistinct.arrayDistinct      array_distinct  avgt   20  26.472 ± 0.411  ns/op

JsonUtil#HashTable, caller use separate blockBuilder
BenchmarkArrayDistinct.arrayDistinct      array_distinct  avgt   20  27.062 ± 0.794  ns/op

Based on the above, I think the best way is to use approach 1), ie. Make JsonUtil#HashTable  a top level class but make it a) be able to handle nulls correctly b) be able to add > 1 blocks. This way we can remove the 4MB limit which users complaint about. MultimapAggregationFunction can still use original TypedSet if we observe the memory pressure. I can make this work for array_intersect first and watch for any performance penalty for adding > 1 blocks.  @dain @wenleix @haozhun what do you think?

wenleix commented:
The benchmark result at top looks great. Note the bigger input arrays are, the better improvements we will see. Thus can we look at the production workload, and generate benchmark with input array of that size ? Given the stack depth of the frame-chart , I suspect the speed-up is likely closer to the upper side in your benchmark.

wenleix commented:
Given the complicated trade-offs between different proposals on `TypedSet`, I would suggest taking one step back and focusing on impact  -- switch from sort based solution to a hash based solution.  

The blocking issue (https://github.com/prestodb/presto/pull/11984#pullrequestreview-178966684) is to avoid new `BlockBuilder` created by `TypedSet` per each invocation. We can just change `TypedSet` to allow a provided container `BlockBuilder` to achieve this goal. All the follow-up optimizations and proposals are great, but let's focus on impact and get rid of sort-based `array_intersect` first ;) . Does that make sense? @haozhun , @yingsu00 

Also note, for heavy use cases, we can always use specialized implementation (For example, we can use `LongSet` instead of `TypedSet` for `BIGINT`)

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:
```diff
@@ -98,26 +106,35 @@ public void add(Block block, int position)
         return blockPositionByHash.get(getHashPositionOfElement(block, position));
     }
 
+    public Block getBlock()
```
I thought it would be better to encapsulate the logic in TypedSet in this method. But if we want to avoid changes to TypedSet as much as possible then I'll remove it. Otherwise I think it makes sense for TypedSet to return the built Block, because caller has passed in the blockBuilder to TypedSet and it's TypedSet's responsibility to return a Block from this blockBuilder.

wenleix replied:
That's a good point. The name `getBlock` is a bit confusing though, since it makes me think it returns the underlying data container block (instead of the region block).

Let's keep the current solution (i.e. do the slice outside `TypedSet`) for now.

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:
```diff
@@ -55,12 +56,19 @@
         this.functionName = functionName;
 
+        elementBlockStartOffset = elementBlock.getPositionCount();
+
```
how about naming it as elementBlockStartSizeInBytes to match with elementBlockStartOffset?

yingsu00 replied:
I renamed them to initialElementBlockOffset and initialElementBlockSizeInBytes

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java:
```diff
@@ -1108,29 +1108,120 @@ public void testArraysOverlap()
                 "ARRAY_INTERSECT(ARRAY [2.3, 2.3, 2.2], ARRAY[2.2, 2.3])",
                 new ArrayType(createDecimalType(2, 1)),
-                ImmutableList.of(decimal("2.2"), decimal("2.3")));
+                ImmutableList.of(decimal("2.3"), decimal("2.2")));
```
Yes. The sort based solution has results in ascending order. The TypedSet result has the same order as how the elements are added.

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArrayIntersect.java:
```diff
@@ -0,0 +1,191 @@
+
+            ImmutableList<RowExpression> projections = projectionsBuilder.build();
+            pageProcessor = compiler.compilePageProcessor(Optional.empty(), projections).get();
+            page = new Page(blocks);
```
The line will be too long. I think it's easier to read just 
            ```
            Block[] blocks = new Block[2];
            blocks[0] = createChannel(POSITIONS, Integer.parseInt(arraySizeString), elementType);
            blocks[1] = createChannel(POSITIONS, Integer.parseInt(arraySizeString), elementType);
            page = new Page(blocks);
```

yingsu00 replied:
I actually changed it to the way you described.

wenleix replied:
Yeah, looks like it's not so long after getting rid of `Integer.parseInt` :)

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java:
```diff
@@ -1108,29 +1108,120 @@ public void testArraysOverlap()
     @Test
     public void testArrayIntersect()
```
ok. Reduced the number of tests. Missing tests were re-added

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:
```diff
@@ -98,26 +106,35 @@ public void add(Block block, int position)
             }
         }
+
+        // null is added to both elementBlock and the blockPositionByHash hash table only once.
```
I think the point is not the "only once" part, but previously null was not added to either blockPositionByHash or elementBlock. Now we are adding null to these.

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestTypedSet.java:
```diff
@@ -67,40 +69,135 @@ public void testConstructor()
+    }
+
+    @Test
+    public void testGetElementPositionWithPassedInSlicedBlockBuilderWithNull()
```
ok, changed the name to testGetElementPositionWithPassedInNonEmptyBlockBuilder

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArrayIntersect.java:
```diff
@@ -0,0 +1,195 @@
+                        elementType.writeBoolean(entryBuilder, ThreadLocalRandom.current().nextBoolean());
+                    }
+                    else if (elementType.equals(VARCHAR)) {
+                        // make sure the size of a varchar is rather small; otherwise the aggregated slice may overflow
```
Each string is at most |arraySize| length. If arraySize = 100, then the strings created would be "0" to "99". It is made this way to avoid the 4MB limit in TypedSet elementBlock to be reached. If the string is very long and all distinct, the 4MB limit could be reached.

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:
```diff
@@ -152,7 +164,7 @@ private void addNewElement(int hashPosition, Block block, int position)
 
         // increase capacity, if necessary
-        if (elementBlock.getPositionCount() >= maxFill) {
+        if (++size >= maxFill) {
```
I put it right above if (size >= maxFill) and after adding the new element to both the elementBlock and blockPositionByHash
